File: cpp/binarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

int binarySearch(vector<int> &nums, int target, int s, int e) {
  if (s > e) {
    return -1;
  }
  int mid = s + (e - s) / 2;
  if (nums[mid] > target) {
    return binarySearch(nums, target, s, mid - 1);
  } else if (nums[mid] < target) {
    return binarySearch(nums, target, mid + 1, e);
  } else {
    return mid;
  }
}

int main() {
  vector<int> nums = {-1, 0, 3, 5, 7, 9, 12};
  int target = -2;
  int start = 0;
  int end = nums.size() - 1;

  cout << binarySearch(nums, target, 0, end);

  return 0;
}

File: python/recursion/binarySearch.py
def recBs(nums: list[int], target, low, high):
    if low > high:
        return -1
    mid = low + (high - low) // 2
    if nums[mid] == target:
        return mid
    elif nums[mid] > target:
        return recBs(nums, target, low, mid - 1)
    else:
        return recBs(nums, target, mid + 1, high)


def bs(nums: list[int], target):
    low = 0
    high = len(nums) - 1
    while low <= high:
        mid = low + (high - low) // 2
        if nums[mid] < target:
            low = mid + 1
        elif nums[mid] > target:
            high = mid - 1
        else:
            return mid
    return -1


nums = [-1, 0, 3, 5, 7, 9]
target = 3
low = 0
high = len(nums) - 1
print(recBs(nums, target, low, high))
